Add lifecycle document (#1274)
dotnet/dotnet-buildtools-prereqs-docker: lifecycle.md

# Prereq container image lifecycle

The container images in this repository are provided for building .NET source and testing the resultant binaries. The images are built with OS base images and must be updated per the OS lifecycle. This document describes the required patterns.

> Container images are produced for each of the supported Linux distros listed in [Supported OS Versions](https://github.com/dotnet/core/blob/main/os-lifecycle-policy.md) files.

Reviewer: They are produced on an as needed basis.  Not all supported distro's especially the cross product of supported architectures will exist unless there is a need for them.

Author: I think the truth is in-between. Ignoring architectures, which distros do you see that are skipped? It appears to me that we've conflated SUSE and OpenSUSE into one, but otherwise all distros are represented. I can change the wording, however, it is good to get on the same page first.

Reviewer: `which distros do you see that are skipped` - Red Hat.

The similar statement in the test section will also need updating:

`Test images are provided for each of the supported OS versions. They are built for each supported architecture.`

Author: We have CentOS Stream and Alma Linux images to cover for Red Hat support.

That's covered by this statement: https://github.com/dotnet/core/blob/main/support.md#compatible-derivatives. I can add a link to that.

Author: I'll address the supported architectures. As you suggest, we only build for all architectures in some cases.

## Build images

Examples:

- [dotnet-buildtools-prereqs-docker #1263](https://github.com/dotnet/dotnet-buildtools-prereqs-docker/pull/1263)
- [dotnet/runtime #110198](https://github.com/dotnet/runtime/pull/110198)

Build images are exclusively built with Azure Linux, supporting both `glibc` and `musl` build flavors. They are built for x64 only, per [Linux build methodology](https://github.com/dotnet/runtime/blob/main/docs/project/linux-build-methodology.md).

Build images are uniquely created for each new .NET version with the latest Azure Linux available at that time. The image tags include both the given .NET and Azure Linux versions.

The Azure Linux lifecycle is shorter than the .NET LTS lifecycle. It is expected that build images will need to be (initially) created with Azure Linux n and then replaced with Azure Linux n+1 mid-way through the .NET version lifecycle.

> Build images are deleted after a .NET version is no longer supported. It is possible that images will  be required for a while after the published EOL date. 

Author: Good point. Resolved.

# Prereq container image lifecycle

The container images in this repository are provided for building .NET source and testing the resultant binaries. The images are built with OS base images and must be updated per the OS lifecycle. This document describes the required patterns.

The images are a mix of Microsoft- and community-supported images.

## Build images

> The build images are exclusively built with Azure Linux, supporting both `glibc` and `musl` build flavors. They are built for x64 only, per [Linux build methodology](https://github.com/dotnet/runtime/blob/main/docs/project/linux-build-methodology.md). Microsoft designates the set of images it used for its official build per [The Official Runtime Docker Images](https://github.com/dotnet/runtime/blob/main/docs/workflow/using-docker.md#the-official-runtime-docker-images).

Author: Addressed. Works?

```suggestion
The build images are primarily built with Azure Linux, supporting both `glibc` and `musl` build flavors. They are built for x64 only, per [Linux build methodology](https://github.com/dotnet/runtime/blob/main/docs/project/linux-build-methodology.md). Microsoft designates the set of images it used for its official build per [The Official Runtime Docker Images](https://github.com/dotnet/runtime/blob/main/docs/workflow/using-docker.md#the-official-runtime-docker-images).
```

> Build images are uniquely created for each new .NET version with the latest Azure Linux available at that time. The image tags include both the given .NET and Azure Linux versions. The tag may also contain a target OS and architecture.

Reviewer 2: This also does not hold for community images today. Maybe this whole chapter should be about `Microsoft official build images" only.

Author: Happy to. If we do that, we should add some text on the community images, too.

I got a bit confused because FreeBSD seems to be fully integrated: https://github.com/dotnet/dotnet-buildtools-prereqs-docker/pull/1272/files.

The Azure Linux lifecycle is shorter than the .NET LTS lifecycle. It is expected that build images will need to be (initially) created with Azure Linux n and then replaced with Azure Linux n+1 mid-way through the .NET version lifecycle. They may also need to be updated if the target OS is EOL.

Official build images are deleted after a .NET version is no longer supported. It is possible that images will be required for a while after the published EOL date.

Examples:

- [dotnet-buildtools-prereqs-docker #1263](https://github.com/dotnet/dotnet-buildtools-prereqs-docker/pull/1263)
- [dotnet/runtime #110198](https://github.com/dotnet/runtime/pull/110198)
- [dotnet-buildtools-prereqs-docker #1272](https://github.com/dotnet/dotnet-buildtools-prereqs-docker/pull/1272)

## Test images

> Test images are produced for the supported Linux distros listed in [Supported OS Versions](https://github.com/dotnet/core/blob/main/os-lifecycle-policy.md) files. In some cases, a [compatible derivative](https://github.com/dotnet/core/blob/main/support.md#compatible-derivatives) will be used. Images are typically built for just `x64` or for all of `Arm32`, `Arm64`, and `x64`. They are shared across .NET versions.

Reviewer 3: > Images are typically built for just `x64` or for all of `Arm32`, `Arm64`, and `x64`.

Phrasing is a little awkward. I might just leave it at, we provide a subset of dependent architectures depending on distribution.

Author: Fair enough. Works?

```suggestion
Test images are produced for the supported Linux distros listed in [Supported OS Versions](https://github.com/dotnet/core/blob/main/os-lifecycle-policy.md) files. In some cases, a [compatible derivative](https://github.com/dotnet/core/blob/main/support.md#compatible-derivatives) will be used. Images are built for a variety of architectures based on need. They are shared across .NET versions.
```

We use version specific references in infra to ensure that our CI builds are reliable. One can imagine using floating OS tags (such as `debian-oldest` and `debian-latest`), however such an approach would be guaranteed to break our build. We know that since we often see build and test breaks that need addressing in PRs where we update build and test images.

At times, it may be necessary to use a [fixed image reference](https://github.com/dotnet/runtime/pull/110199#discussion_r1859075989) for build reliability. If this is ever done, a tracking issue should be created (before the PR is merged) so that we remember to resolve the underlying issue and update the image reference.

The following locations are examples of infra that gets updated when new images are available.

- https://github.com/dotnet/runtime/blob/main/eng/pipelines/libraries/helix-queues-setup.yml
- https://github.com/dotnet/runtime/blob/main/eng/pipelines/coreclr/templates/helix-queues-setup.yml
- https://github.com/dotnet/arcade/tree/main/eng/common/templates-official/

> - https://github.com/dotnet/aspnetcore/tree/main/.azure/pipelines

Reviewer 4: The linter is complaining because these are bare URLs. Is this resolved by enclosing them in angle brackets: `<url>`?

# Prereq container image lifecycle

The container images in this repository are provided for building .NET source and testing the resultant binaries. The images are built with OS base images and must be updated per the OS lifecycle. This document describes the required patterns.

The images are a mix of Microsoft- and community-supported images.

## Microsoft build images

> The build images are primarily built with Azure Linux, supporting both `glibc` and `musl` build flavors. They are built for x64 only, per [Linux build methodology](https://github.com/dotnet/runtime/blob/main/docs/project/linux-build-methodology.md). Microsoft designates the set of images it used for its official build per [The Official Runtime Docker Images](https://github.com/dotnet/runtime/blob/main/docs/workflow/using-docker.md#the-official-runtime-docker-images).

Reviewer 2: ```suggestion
The build images are built with Azure Linux, supporting both `glibc` and `musl` build flavors. They are built for x64 only, per [Linux build methodology](https://github.com/dotnet/runtime/blob/main/docs/project/linux-build-methodology.md). Microsoft designates the set of images it used for its official build per [The Official Runtime Docker Images](https://github.com/dotnet/runtime/blob/main/docs/workflow/using-docker.md#the-official-runtime-docker-images).
```
All Microsoft build images are Azure Linux
